### PR TITLE
hamlib: 3.3 -> 4.2

### PR DESCRIPTION
**Pull request by hexagonal-sun**
###### Motivation for this change

Use latest version.  Also includes:

 - Remove depricated configure flag `--with-rigmatrix`
 - Made various bindings configurable.
 - Set python bindings to `false` by default as I couldn't get this to build.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [x] macOS (apple silicon)
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/libraries/hamlib/default.nix [hunk 1]
```diff
@@ -1,20 +1,46 @@
...
+, libxml2
+, tcl
+, libusb-compat-0_1
+, pkg-config
+, boost
+, libtool
+, perlPackages
+, pythonBindings ? false
+, tclBindings ? true
+, perlBindings ? true }:
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/hamlib/default.nix:17]:
```suggestion
, perlBindings ? true
}:
```

File: pkgs/development/libraries/hamlib/default.nix [hunk 2]
```diff
@@ -1,20 +1,46 @@
...
+    gd
+    libxml2
+    libusb-compat-0_1
+    pkg-config
+    boost
+    libtool
+  ] ++ lib.optionals pythonBindings [ python3 ]
+    ++ lib.optionals tclBindings [ tcl ]
+    ++ lib.optionals perlBindings [ perl
+                                    perlPackages.ExtUtilsMakeMaker ];
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/hamlib/default.nix:39]:
```suggestion
    ++ lib.optionals perlBindings [ perl perlPackages.ExtUtilsMakeMaker ];
```

File: pkgs/development/libraries/hamlib/default.nix [hunk 3]
```diff
@@ -1,20 +1,46 @@
...
 
-  configureFlags = [ "--with-perl-binding" "--with-python-binding"
-                     "--with-tcl-binding" "--with-rigmatrix" ];
+  configureFlags = lib.optionals perlBindings [ "--with-perl-binding" ]
+                ++ lib.optionals tclBindings [ "--with-tcl-binding" "--with-tcl=${tcl}/lib/" ]
+                ++ lib.optionals pythonBindings [ "--with-python-binding" ];
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/hamlib/default.nix:43]:
```suggestion
  configureFlags = lib.optionals perlBindings [ "--with-perl-binding" ]
    ++ lib.optionals tclBindings [ "--with-tcl-binding" "--with-tcl=${tcl}/lib/" ]
    ++ lib.optionals pythonBindings [ "--with-python-binding" ];
```

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 129950` at 1b6f1a06 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages marked as broken and skipped:</summary>
  <ul>
    <li>gnuradio3_7Packages.ais</li>
    <li>gnuradio3_7Packages.gsm</li>
    <li>gnuradio3_7Packages.osmosdr</li>
  </ul>
</details>
<details>
  <summary>15 packages failed to build:</summary>
  <ul>
    <li><a href="https://gist.github.com/a97c70ac478cdbf761ea0a0c5e07d45e">cubicsdr</a></li>
    <li><a href="https://gist.github.com/2339c5530a8fb5318395caa5f4059512">gnss-sdr</a></li>
    <li><a href="https://gist.github.com/57a20aa2cd6eb8214c4433688545b584">gnuradio3_8Packages.ais</a></li>
    <li><a href="https://gist.github.com/5161558a70ce248c0e1ca6ea2d7433af">gnuradio3_8Packages.osmosdr</a></li>
    <li><a href="https://gist.github.com/6fb58f304b3e040d98638c0579b7d2a4">gqrx</a></li>
    <li><a href="https://gist.github.com/77a4a18d000fda7f59434ca249dc5664">klog</a></li>
    <li><a href="https://gist.github.com/cf04cc02af21320a3f1882895872366c">python38Packages.soapysdr-with-plugins</a></li>
    <li><a href="https://gist.github.com/30d5b19defd43b50c6fdc305558e9716">python39Packages.soapysdr-with-plugins</a></li>
    <li><a href="https://gist.github.com/cb47b9cafe175a56d27190daec1bd7aa">qsstv</a></li>
    <li><a href="https://gist.github.com/bd750e8240d36114797b900c837d49c8">rtl_433</a></li>
    <li><a href="https://gist.github.com/50ef182da51acec8079aceba5fde311b">soapyaudio</a></li>
    <li><a href="https://gist.github.com/16fd169e0df13c6af5760acdf6a558e3">soapysdr-with-plugins</a></li>
    <li><a href="https://gist.github.com/502f84b51ed91ea3e862c9366b2bba10">tlf</a></li>
    <li><a href="https://gist.github.com/05ba4e5b99aafdf5cdaa719d1e49dee3">welle-io</a></li>
    <li><a href="https://gist.github.com/a40303a523b99c4af1c27d49608c597c">xlog</a></li>
  </ul>
</details>
<details>
  <summary>3 packages skipped due to time constraints:</summary>
  <ul>
    <li>pothos</li>
    <li>qradiolink</li>
    <li>sdrangel</li>
  </ul>
</details>
<details>
  <summary>5 packages built successfully:</summary>
  <ul>
    <li>cqrlog</li>
    <li>direwolf</li>
    <li>fldigi</li>
    <li>gpredict</li>
    <li>hamlib</li>
  </ul>
</details>
<details>
  <summary>3 suggestions:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/build-tools-in-build-inputs.md">build-tools-in-build-inputs</a>

swig is a build tool so it likely goes to `nativeBuildInputs`, not `buildInputs`.

Near pkgs/development/libraries/hamlib/default.nix:28:3:
```
   |
28 |   buildInputs = [
   |   ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/build-tools-in-build-inputs.md">build-tools-in-build-inputs</a>

libtool is a build tool so it likely goes to `nativeBuildInputs`, not `buildInputs`.

Near pkgs/development/libraries/hamlib/default.nix:28:3:
```
   |
28 |   buildInputs = [
   |   ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/build-tools-in-build-inputs.md">build-tools-in-build-inputs</a>

pkg-config is a build tool so it likely goes to `nativeBuildInputs`, not `buildInputs`.

Near pkgs/development/libraries/hamlib/default.nix:28:3:
```
   |
28 |   buildInputs = [
   |   ^
```
</li>
  </ul>
</details>

Note that build failures [may predate](https://github.com/nix-community/hydra-check) this PR, and could be nondeterministic or hardware dependent.
Please exercise your independent judgement. Does something look off? Please file an [issue](https://github.com/rmcgibbo/r-rmcgibbo/issues) or reach out on IRC.

**Comment by hexagonal-sun** [comment]
@SuperSandro2000 Thanks for reviewing.  Comments addressed.

**Comment by hexagonal-sun** [comment]
Fixed `nativeBuildInputs` as per @r-rmcgibbo's suggestions.  I got the python bindings build working on apple silicon, so I've re-enabled it.  It turns out the `gettext` package is implicitly required when building python bindings.

File: pkgs/development/libraries/hamlib/default.nix [hunk 4]
```diff
@@ -1,20 +1,50 @@
-{lib, stdenv, fetchurl, perl, python2, swig, gd, libxml2, tcl, libusb-compat-0_1, pkg-config,
- boost, libtool, perlPackages }:
+{ lib
+, stdenv
+, fetchurl
+, perl
+, swig
+, gd
+, python2
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/hamlib/default.nix:7]: Can we use python3 instead?

Comment (hexagonal-sun) [reply pkgs/development/libraries/hamlib/default.nix]: Indeed, I've managed to get this building with `python3`.  Changes pushed.

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).

Result of `nixpkgs-review pr 129950` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
3 packages marked as broken and skipped:
</summary>
<ul>
<li>gnuradio3_7Packages.ais</li>
<li>gnuradio3_7Packages.gsm</li>
<li>gnuradio3_7Packages.osmosdr</li>
</ul>
</details>
<details>

  <summary>18 packages failed to build and are new build failures:</summary>
  <ul>
<li>cubicsdr: log was empty</li>
<li>gnss-sdr: log was empty</li>
<li>gnuradio3_8Packages.ais: log was empty</li>
<li>gnuradio3_8Packages.osmosdr: log was empty</li>
<li>gqrx: log was empty</li>
<li>klog: log was empty</li>
<li>pothos: log was empty</li>
<li>python38Packages.soapysdr-with-plugins: log was empty</li>
<li>python39Packages.soapysdr-with-plugins: log was empty</li>
<li>qradiolink: log was empty</li>
<li>qsstv: log was empty</li>
<li>rtl_433: log was empty</li>
<li>sdrangel: log was empty</li>
<li>soapyaudio: log was empty</li>
<li>soapysdr-with-plugins: log was empty</li>
<li>tlf: log was empty</li>
<li>welle-io: log was empty</li>
<li>xlog: log was empty</li>
  </ul>
</details>
<details>
<summary>
5 packages built:
</summary>
<ul>
<li>cqrlog</li>
<li>direwolf</li>
<li>fldigi</li>
<li>gpredict</li>
<li>hamlib</li>
</ul>
</details>

**Comment by risicle** [comment]
This version of `hamlib` does at least build successfully on darwin (macos 10.15).

Though yes, many reverse dependencies broken on linux.

**Comment by SuperSandro2000** [comment]
I think we need to keep hamlib that points to hamblib_3 around here and add a new hamlib_4.

**Comment by hexagonal-sun** [comment]
@SuperSandro2000 Sounds like a reasonable approach.  Then presumably anything that is compatible with hamlib_4 could be uplifted, and everything else still uses hamlib_3?

**Comment by SuperSandro2000** [comment]
Ys, exactly.

**Comment by hexagonal-sun** [comment]
@SuperSandro2000 Okay, I've split this out into two parts and uplifed the modules that built successfully against hamlib 4.2

File: pkgs/top-level/all-packages.nix [hunk 5]
```diff
@@ -15692,7 +15692,9 @@ in
 
   gwenhywfar = callPackage ../development/libraries/aqbanking/gwenhywfar.nix { };
 
-  hamlib = callPackage ../development/libraries/hamlib { };
+  hamlib = hamlib_3;
+  hamlib_3 = callPackage ../development/libraries/hamlib { };
+  hamlib_4 = callPackage ../development/libraries/hamlib/4.2.nix { };
```
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:15697]:
```suggestion
  hamlib_4 = callPackage ../development/libraries/hamlib/4.nix { };
```

File: pkgs/development/libraries/hamlib/4.2.nix [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    can use to send the appropriate commands to a radio.
+
+    Also included in the package is a simple radio control program 'rigctl',
+    which lets one control a radio transceiver or receiver, either from
+    command line interface or in a text-oriented interactive interface.
+    '';
+    license = with lib.licenses; [ gpl2Plus lgpl2Plus ];
+    homepage = "http://hamlib.sourceforge.net";
+    maintainers = with lib.maintainers; [ relrod ];
+    platforms = with lib.platforms; unix;
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/hamlib/4.2.nix:62]:
```suggestion
  meta = with lib; {
    description = "Runtime library to control radio transceivers and receivers";
    longDescription = ''
    Hamlib provides a standardized programming interface that applications
    can use to send the appropriate commands to a radio.

    Also included in the package is a simple radio control program 'rigctl',
    which lets one control a radio transceiver or receiver, either from
    command line interface or in a text-oriented interactive interface.
    '';
    license = with licenses; [ gpl2Plus lgpl2Plus ];
    homepage = "http://hamlib.sourceforge.net";
    maintainers = with maintainers; [ relrod ];
    platforms = platforms.unix;
```

File: pkgs/development/libraries/hamlib/default.nix [hunk 7]
```diff
@@ -1,20 +1,50 @@
...
+}:
 
 stdenv.mkDerivation rec {
   pname = "hamlib";
   version = "3.3";
 
   src = fetchurl {
     url = "mirror://sourceforge/${pname}/${pname}-${version}.tar.gz";
-    sha256 = "10788mgrhbc57zpzakcxv5aqnr2819pcshml6fbh8zvnkja562y9";
+    sha256 =  "10788mgrhbc57zpzakcxv5aqnr2819pcshml6fbh8zvnkja562y9";
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/hamlib/default.nix:27]:
```suggestion
    sha256 = "10788mgrhbc57zpzakcxv5aqnr2819pcshml6fbh8zvnkja562y9";
```

**Review by SuperSandro2000 — COMMENTED** [review]
Please overwrite the hamlib in inputs isntead of replacing it everywhere.

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).

Result of `nixpkgs-review pr 129950` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
3 packages marked as broken and skipped:
</summary>
<ul>
<li>gnuradio3_7Packages.ais</li>
<li>gnuradio3_7Packages.gsm</li>
<li>gnuradio3_7Packages.osmosdr</li>
</ul>
</details>
<details>
<summary>
24 packages built:
</summary>
<ul>
<li>cqrlog</li>
<li>cubicsdr</li>
<li>direwolf</li>
<li>fldigi</li>
<li>gnss-sdr</li>
<li>gnuradio3_8Packages.ais</li>
<li>gnuradio3_8Packages.osmosdr</li>
<li>gpredict</li>
<li>gqrx</li>
<li>hamlib (hamlib_3)</li>
<li>hamlib_4</li>
<li>klog</li>
<li>pothos</li>
<li>python38Packages.soapysdr-with-plugins</li>
<li>python39Packages.soapysdr-with-plugins</li>
<li>qradiolink</li>
<li>qsstv</li>
<li>rtl_433</li>
<li>sdrangel</li>
<li>soapyaudio</li>
<li>soapysdr-with-plugins</li>
<li>tlf</li>
<li>welle-io</li>
<li>xlog</li>
</ul>
</details>

**Comment by hexagonal-sun** [comment]
> Please overwrite the hamlib in inputs isntead of replacing it everywhere.

Ah yes, that's neater.  Done.

**Pull request merged**